### ISSUE TITLE
TICCL-LDcalc produces duplicate output lines, omitting others....

### ISSUE THREAD

**Issue by kosloot** [LanguageMachines/ticcltools#13]
occasionally TICCL-rank generates the same line twice:
```
 Noordelyk#1#noodlyk#100000000#2#0.965138
 Noordelyk#1#noodlyk#100000000#2#0.965138
```

and:
```
  terwy#1#werry#100000000#2#0.995248
! terwy#1#werry#100000000#2#0.995248
  terwyi#1#teremi#100000000#2#0.957082
--- 9753,9755 ----
  terwy#1#werry#100000000#2#0.995248
! terwy#1#Werry#100000000#2#0.995248
  terwyi#1#teremi#100000000#2#0.957082
```
where ``terwy#1#werry#100000000#2#0.995248`` is doubled, 
and ``terwy#1#Werry#100000000#2#0.995248`` is missed...

Maybe a 1-0ff error construction the threads?

**Comment by kosloot** [comment]
Ok, this might already be a problem in LDcalc.

``
Noordelyk~1~1~noodlyk~100000000~100000000~23207337056~2~7~1~1~1~0
`` 
is found twice there, as is
``
terwy~1~1~werry~100000000~200000000~596286601~2~3~1~0~0~0
``

**Comment by kosloot** [comment]
I suspect a threading problem here, but even with -t 1 results are not deterministic.
Have to look into that

**Comment by kosloot** [comment]
it seems to be a problem related to LDcalc's handling of .indexNT files vs .index files.
Even if these files are identical, small differences occur in the output.

investigation continues....

**Comment by kosloot** [comment]
In fact this is INTENDED. On files generated by TICCL-indexerNT,  LDcalc also looks **backward**; so on exactly the same file, extra results are to be expected. 
The **backward** search was especially introduced to work on indexerNT files. The files generated by TICCL-indexer were originally a lot different (larger). But when used with the same 'foci' file, TICCL-indexer and TICCL-indexerNT generate exactly the same files. Is that even correct?
Maybe TICCLindexer should generate more indexes, to compensate for the missing backward search?

O, this is so scary..

**Comment by martinreynaert** [comment]
How Indexer and IndexerNT (should) work

We can now have three input files for both tools:
- corpus word anagram hash: one value from this = CWAV
- focus word anagram values list: one value from this = FWAV
- character confusion anagram values list (up to a particular LD). One value from this = CCAV.

Indexer takes each CCAV in turn and runs over the CWAV list, adding up both values. If the sum of the two (necessarily a higher number and the CWAV list is sorted ascendingly) gives a hit, the indexes to the pair of words displaying the particular confusion represented by the CCAV have been indentified and need to be reported. If run without focus word list, all these hits are reported. If run with a focus list, only those hits involving CWAVs that occur in the FWAV list, are reported.

IndexerNT takes a value from the FWAV list and calculates the numerical difference with all the CWAVs surrounding this value in the CWAV list within the bounds of the uppermost value (the last, this list too being ascendingly sorted) of the CCAV list preceding (= the backward look) or following the FWAV in the corpus anagram list. Any absolute numerical difference between CWAV and the FWAV that is also present in the CCAV list, signals a pair that should be reported.

**Comment by kosloot** [comment]
This is somewhat clear.
The indexer _with_ FWAV produces the same files as indexerNT.
And LDcalc should treat them equally.  let's call it **CASE 1**
The indexer _without__ FWAV produces a much larger file that is handled differently **CASE**2

At the moment, LDcalc handles **CASE  1** like **CASE 2**, but additionally does a backward search. This introduces doubles, but also other variants, which may or may be not correct.

Probably, both cases can be folded to one , when we store the correct indexes in indexer and indexerNT.
As @martinreynaert stated:
``het is dus een kwestie van hoe je de paren opslaat. Als je bij een CCAV altijd de laagste van het daardoor gelinkte CWAV paar opslaat, dan moet LDcalc verder altijd alleen maar vooruit kijken``

So it is time to investigate indexer again.

**Comment by kosloot** [comment]
OK, so I ran indexer on the DPO file, creating a large .index file:
``ls -alh  OUT/book/TICCL/TESTDP035.tsv.clean.index 
-rw-r--r-- 1 sloot sloot 1.8G Mar  6 15:26 OUT/book/TICCL/TESTDP035.tsv.clean.index
``

compare this to the indexerNT variant:
``ls -alh  OUT/book/TICCL/TESTDP035.tsv.clean.indexNT 
-rw-r--r-- 1 sloot sloot 19M Mar  6 14:25 OUT/book/TICCL/TESTDP035.tsv.clean.indexNT
``

Then I ran LDcalc on both files, creating:
``wc T.*
  121030   121030  7511226 T.ld
  122320   122320  7589314 T.NT.ld
``

the .NT. variant is just a little bit larger.
Then I ran LDcalc on the indexerNT **WITHOUT** backward search.
This produces then __exactly__ the same .ldcalc file as on the 1.9 inputfile.

Conclusion: The backward option is not longer needed, and even harmful!.
We could remove it from LDcalc, making LDcalc opaque for indexer versus indexerNT files AND
making the difference between .index and .indexNT files obsolete too.

A major simplification!
